File: packages/flatsql-tools/src/grammar/enums.rs
use flatsql_proto as proto;

pub fn get_enum_text(target: &proto::Node) -> &'static str {
    let v = target.children_begin_or_value();
    match target.node_type() {
        proto::NodeType::ENUM_SQL_GROUP_BY_ITEM_TYPE => {
            proto::GroupByItemType(v as u8).variant_name().unwrap_or_default()
        }
        proto::NodeType::ENUM_SQL_TEMP_TYPE => proto::TempType(v as u8).variant_name().unwrap_or_default(),
        proto::NodeType::ENUM_SQL_CONST_TYPE => proto::AConstType(v as u8).variant_name().unwrap_or_default(),
        proto::NodeType::ENUM_SQL_CHARACTER_TYPE => proto::CharacterType(v as u8).variant_name().unwrap_or_default(),
        proto::NodeType::ENUM_SQL_EXPRESSION_OPERATOR => {
            proto::ExpressionOperator(v as u8).variant_name().unwrap_or_default()
        }
        proto::NodeType::ENUM_SQL_ORDER_DIRECTION => proto::OrderDirection(v as u8).variant_name().unwrap_or_default(),
        proto::NodeType::ENUM_SQL_ORDER_NULL_RULE => proto::OrderNullRule(v as u8).variant_name().unwrap_or_default(),
        proto::NodeType::ENUM_SQL_COMBINE_MODIFIER => {
            proto::CombineModifier(v as u8).variant_name().unwrap_or_default()
        }
        proto::NodeType::ENUM_SQL_COMBINE_OPERATION => {
            proto::CombineOperation(v as u8).variant_name().unwrap_or_default()
        }
        proto::NodeType::ENUM_SQL_NUMERIC_TYPE => proto::NumericType(v as u8).variant_name().unwrap_or_default(),
        proto::NodeType::ENUM_SQL_WINDOW_BOUND_MODE => {
            proto::WindowBoundMode(v as u8).variant_name().unwrap_or_default()
        }
        proto::NodeType::ENUM_SQL_WINDOW_RANGE_MODE => {
            proto::WindowRangeMode(v as u8).variant_name().unwrap_or_default()
        }
        proto::NodeType::ENUM_SQL_WINDOW_EXCLUSION_MODE => {
            proto::WindowExclusionMode(v as u8).variant_name().unwrap_or_default()
        }
        proto::NodeType::ENUM_SQL_WINDOW_BOUND_DIRECTION => {
            proto::WindowBoundDirection(v as u8).variant_name().unwrap_or_default()
        }
        proto::NodeType::ENUM_SQL_ON_COMMIT_OPTION => proto::OnCommitOption(v as u8).variant_name().unwrap_or_default(),
        proto::NodeType::ENUM_SQL_CONSTRAINT_ATTRIBUTE => {
            proto::ConstraintAttribute(v as u8).variant_name().unwrap_or_default()
        }
        proto::NodeType::ENUM_SQL_COLUMN_CONSTRAINT => {
            proto::ColumnConstraint(v as u8).variant_name().unwrap_or_default()
        }
        proto::NodeType::ENUM_SQL_INTERVAL_TYPE => proto::IntervalType(v as u8).variant_name().unwrap_or_default(),
        proto::NodeType::ENUM_SQL_SUBQUERY_QUANTIFIER => {
            proto::SubqueryQuantifier(v as u8).variant_name().unwrap_or_default()
        }
        proto::NodeType::ENUM_SQL_KNOWN_FUNCTION => proto::KnownFunction(v as u8).variant_name().unwrap_or_default(),
        proto::NodeType::ENUM_SQL_TRIM_TARGET => proto::TrimDirection(v as u8).variant_name().unwrap_or_default(),
        proto::NodeType::ENUM_SQL_EXTRACT_TARGET => proto::ExtractTarget(v as u8).variant_name().unwrap_or_default(),
        proto::NodeType::ENUM_SQL_ROW_LOCKING_BLOCK_BEHAVIOR => proto::RowLockingBlockBehavior(v as u8)
            .variant_name()
            .unwrap_or_default(),
        proto::NodeType::ENUM_SQL_ROW_LOCKING_STRENGTH => {
            proto::RowLockingStrength(v as u8).variant_name().unwrap_or_default()
        }
        proto::NodeType::ENUM_SQL_SAMPLE_UNIT_TYPE => {
            proto::SampleCountUnit(v as u8).variant_name().unwrap_or_default()
        }

        proto::NodeType::ENUM_SQL_JOIN_TYPE => proto::JoinType(v as u8).variant_name().unwrap_or_default(),

        _ => "?",
    }
}

File: packages/sqlynx/src/parser/grammar/enums.cc
#include "sqlynx/parser/grammar/enums.h"

#include <algorithm>

#include "sqlynx/proto/proto_generated.h"

namespace sqlynx {
namespace parser {

const char* getEnumText(const proto::Node& target) {
    auto nt = target.node_type();
    auto v = static_cast<uint32_t>(target.children_begin_or_value());
    switch (nt) {
#define X(ENUM_TYPE, TYPE_TABLE)     \
    case proto::NodeType::ENUM_TYPE: \
        return proto::TYPE_TABLE()->names[v];
        X(ENUM_SQL_CHARACTER_TYPE, CharacterTypeTypeTable)
        X(ENUM_SQL_COLUMN_CONSTRAINT, ColumnConstraintTypeTable)
        X(ENUM_SQL_COMBINE_MODIFIER, CombineModifierTypeTable)
        X(ENUM_SQL_COMBINE_OPERATION, CombineOperationTypeTable)
        X(ENUM_SQL_CONSTRAINT_ATTRIBUTE, ConstraintAttributeTypeTable)
        X(ENUM_SQL_CONST_TYPE, AConstTypeTypeTable)
        X(ENUM_SQL_EXPRESSION_OPERATOR, ExpressionOperatorTypeTable)
        X(ENUM_SQL_EXTRACT_TARGET, ExtractTargetTypeTable)
        X(ENUM_SQL_GROUP_BY_ITEM_TYPE, GroupByItemTypeTypeTable)
        X(ENUM_SQL_INTERVAL_TYPE, IntervalTypeTypeTable)
        X(ENUM_SQL_KNOWN_FUNCTION, KnownFunctionTypeTable)
        X(ENUM_SQL_NUMERIC_TYPE, NumericTypeTypeTable)
        X(ENUM_SQL_ON_COMMIT_OPTION, OnCommitOptionTypeTable)
        X(ENUM_SQL_ORDER_DIRECTION, OrderDirectionTypeTable)
        X(ENUM_SQL_ORDER_NULL_RULE, OrderNullRuleTypeTable)
        X(ENUM_SQL_ROW_LOCKING_BLOCK_BEHAVIOR, RowLockingBlockBehaviorTypeTable)
        X(ENUM_SQL_ROW_LOCKING_STRENGTH, RowLockingStrengthTypeTable)
        X(ENUM_SQL_SUBQUERY_QUANTIFIER, SubqueryQuantifierTypeTable)
        X(ENUM_SQL_TEMP_TYPE, TempTypeTypeTable)
        X(ENUM_SQL_TRIM_TARGET, TrimDirectionTypeTable)
        X(ENUM_SQL_WINDOW_BOUND_DIRECTION, WindowBoundDirectionTypeTable)
        X(ENUM_SQL_WINDOW_BOUND_MODE, WindowBoundModeTypeTable)
        X(ENUM_SQL_WINDOW_EXCLUSION_MODE, WindowExclusionModeTypeTable)
        X(ENUM_SQL_WINDOW_RANGE_MODE, WindowRangeModeTypeTable)
#undef X

        case proto::NodeType::ENUM_SQL_JOIN_TYPE: {
            auto tt = proto::JoinTypeTypeTable();
            auto iter =
                std::lower_bound(tt->values, tt->values + tt->num_elems, v, [](auto l, auto r) { return l < r; });
            if (iter >= (tt->values + tt->num_elems) || *iter != v) {
                return "?";
            }
            auto idx = iter - tt->values;
            return proto::JoinTypeTypeTable()->names[idx];
        }

        default:
            return "?";
    }
}

}  // namespace parser
}  // namespace sqlynx
